src/Dependencies/Collections/SegmentedArray`1.cs

```diff
@@ -94,9 +116,31 @@ private SegmentedArray(int length, T[][] items)
 
         public ref T this[int index]
         {
+            [MethodImpl(SegmentedArrayHelper.FastPathMethodImplOptions)]
             get
             {
-                return ref _items[index >> s_segmentShift][index & s_offsetMask];
+#if NET
+                if (typeof(T).IsValueType)
```

Comment:
📝 Removing this conditional check results in a small improvement to the code generated for the `SegmentedArray<object>` test:

```patch
 ; IdeCoreBenchmarks.SegmentedArrayBenchmarks.ShiftAllSegmentedObject()
        push      rdi
        push      rsi
        push      rbp
        push      rbx
        sub       rsp,28
        mov       rsi,rcx
        xor       edi,edi
        jmp       short M00_L01
 M00_L00:
        mov       rdx,rbx
        mov       rdx,[rdx]
        mov       ecx,ebp
        cmp       edi,ecx
        jae       near ptr M00_L02
        cmp       [rdx],edx
        add       rdx,10
        mov       ecx,edi
        sar       ecx,0D
        movsxd    rcx,ecx
        mov       rdx,[rdx+rcx*8]
        cmp       [rdx],edx
        add       rdx,10
        mov       ecx,edi
        and       ecx,1FFF
        movsxd    rcx,ecx
        lea       rcx,[rdx+rcx*8]
        inc       edi
        mov       edx,edi
        mov       rax,[rbx]
        cmp       edx,ebp
        jae       near ptr M00_L02
        cmp       [rax],eax
        add       rax,10
        mov       r8d,edx
        sar       r8d,0D
        movsxd    r8,r8d
        mov       rax,[rax+r8*8]
        cmp       [rax],eax
        add       rax,10
        and       edx,1FFF
        movsxd    rdx,edx
-       lea       rdx,[rax+rdx*8]
-       mov       rdx,[rdx]
+       mov       rdx,[rax+rdx*8]
        call      CORINFO_HELP_CHECKED_ASSIGN_REF
 M00_L01:
        lea       rbx,[rsi+30]
        mov       rdx,rbx
        mov       ebp,[rdx+8]
        lea       edx,[rbp+0FFFF]
        cmp       edi,edx
        jl        near ptr M00_L00
        mov       rax,rbx
        lea       edx,[rbp+0FFFF]
        mov       rax,[rax]
        mov       ecx,ebp
        cmp       edx,ecx
        jae       short M00_L02
        cmp       [rax],eax
        add       rax,10
        mov       ecx,edx
        sar       ecx,0D
        movsxd    rcx,ecx
        mov       rax,[rax+rcx*8]
        cmp       [rax],eax
        add       rax,10
        and       edx,1FFF
        movsxd    rdx,edx
-       lea       rax,[rax+rdx*8]
-       xor       edx,edx
-       mov       [rax],rdx
+       xor       ecx,ecx
+       mov       [rax+rdx*8],rcx
        add       rsp,28
        pop       rbx
        pop       rbp
        pop       rsi
        pop       rdi
        ret
 M00_L02:
        call      Microsoft.CodeAnalysis.Collections.Internal.ThrowHelper.ThrowIndexOutOfRangeException()
        int       3
-; Total bytes of code 225
+; Total bytes of code 219
```

Reply:
> Removing this conditional check results in a small improvement to the code generated for the SegmentedArray<object> test

Did you measure this on core or netfx?  It should have zero impact on the asm on core, as the JIT now recognizes the pattern.

Reply:
The diff above was from 5.0.3.

Reply:
Actually looking at the diff now... the difference you're seeing is just whether there's an extra `lea`?

Reply:
➡️ Now filed as https://github.com/dotnet/runtime/issues/48605